File: DummyArchive.m
classdef DummyArchive < handle
    % This archive is used only to report found solutions according to the template 
    % required by CEC'2020 competition. The optimization algorithm does not use information
    % from this archive 
    properties 
        solution; % solutions
        value=[] % values
        foundEval=[] % the number of function evaluation according to the algorithmic count
        foundEval2=[] % the number of function evaluation according to the algorithmic count
        foundTime=[] % time when it was found (ms)
        actionCode=[] % actione code (-1, 0, or 1)
    end
    methods
        function dummyArchive=DummyArchive(problem)
            dummyArchive.solution=zeros(0,problem.dim); 
        end  
        function append(dummyArchive,action,index,archive,problem) 
            % Append solutions with proper action code to the dummy archive 
            dummyArchive.solution= [dummyArchive.solution; archive.solution(index,:)]; % add the solution 
            dummyArchive.value=[dummyArchive.value, archive.value(index)]; % add the value 
            dummyArchive.foundEval=[dummyArchive.foundEval, archive.foundEval(index)]; % add the evaluation until found (internal algorithm count of evaluations)
            dummyArchive.foundEval2=[dummyArchive.foundEval2, problem.numCallF]; % add the evaluation until found (internal algorithm count of evaluations)

            dummyArchive.foundTime=[dummyArchive.foundTime, archive.foundTime(index)]; % time (ms) at which it was found
            dummyArchive.actionCode=[dummyArchive.actionCode, action*ones(1,numel(index)) ]; % action code
        end
    end
end

File: dummy_archive.py
"""Dummy archive for CEC competition reporting format."""
from __future__ import annotations
import numpy as np
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .optim_problem import OptimProblem


class DummyArchive:
    """Archive used only to report found solutions according to CEC'2020 template."""
    
    def __init__(self, problem: OptimProblem):
        self.solution = np.zeros((0, problem.dim))
        self.value = np.array([])
        self.foundEval = np.array([])
        self.foundEval2 = np.array([])
        self.foundTime = np.array([])
        self.actionCode = np.array([])

    def append(self, action: int, index: int | list | np.ndarray, archive, problem: OptimProblem):
        """Append solutions with proper action code to the dummy archive."""
        if isinstance(index, int):
            index = [index]
        index = np.asarray(index)
        
        self.solution = np.vstack([self.solution, archive.solution[index, :]])
        self.value = np.concatenate([self.value, archive.value[index]])
        self.foundEval = np.concatenate([self.foundEval, archive.foundEval[index]])
        self.foundEval2 = np.concatenate([self.foundEval2, np.full(len(index), problem.numCallF)])
        self.foundTime = np.concatenate([self.foundTime, archive.foundTime[index]])
        self.actionCode = np.concatenate([self.actionCode, np.full(len(index), action)])
